### PR TITLE
refactor: Remove negative lock annotations from globals

### PR DESCRIPTION
**Pull request by maflcko**
They only make sense for mutexes that are private members. Until cs_main is a private member the negative annotations should be replaced by excluded annotations, which are optional.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK (in the light of #20272 discussion).

**Comment by szeli57** [comment]
2021. április 5., hétfő dátummal Hennadii Stepanov ***@***.***>
ezt írta:

**Review by hebasto — COMMENTED** [review]
Tested faadb1b050dd37ea8d294def554f1da22029d3e6 on Ubuntu 21.10:
```
$ clang --version
Ubuntu clang version 12.0.0-++rc3-4ubuntu1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

With this PR massive TSA warnings are gone.

Could the global namespace be explicitly used for global mutexes in the touched lines?

**Review by hebasto — CHANGES_REQUESTED** [review]
May I suggest to:
- add some missed annotations:
```diff
diff --git a/src/index/base.h b/src/index/base.h
index 8559e3cb6..ac3c429a5 100644
--- a/src/index/base.h
+++ b/src/index/base.h
@@ -109,7 +109,7 @@ public:
     /// sync once and only needs to process blocks in the ValidationInterface
     /// queue. If the index is catching up from far behind, this method does
     /// not block and immediately returns false.
-    bool BlockUntilSyncedToCurrentChain() const;
+    bool BlockUntilSyncedToCurrentChain() const LOCKS_EXCLUDED(::cs_main);
 
     void Interrupt();
 
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index 7932bd291..3a650923c 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -815,7 +815,7 @@ static RPCHelpMan signrawtransactionwithkey()
     };
 }
 
-static RPCHelpMan sendrawtransaction()
+static RPCHelpMan sendrawtransaction() LOCKS_EXCLUDED(::cs_main)
 {
     return RPCHelpMan{"sendrawtransaction",
                 "\nSubmit a raw transaction (serialized, hex-encoded) to local node and network.\n"
```
- fix `EXCLUSIVE_LOCKS_REQUIRED(!::cs_main)` example in https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

?

**Comment by hebasto** [comment]
This change reverts the @ajtowns's [idea](https://github.com/bitcoin/bitcoin/pull/19668#issuecomment-669312538) for `AssertLockNotHeld` but I cannot see the way to save it, unfortunately.

**Comment by ajtowns** [comment]
Approach ACK, +1 on doc change as well

**Comment by MarcoFalke** [comment]
> This change reverts

Correct, but the revert is only temporary (until cs_main is removed or a private member)

**Comment by MarcoFalke** [comment]
Force pushed to address feedback

**Review by hebasto — APPROVED** [review]
ACK fa5eabe72117f6e3704858e8d5b2c57a120258ed

**Review by vasild — APPROVED** [review]
ACK fa5eabe72117f6e3704858e8d5b2c57a120258ed

No warnings with Clang 13.

**Comment by sipa** [comment]
utACK fa5eabe72117f6e3704858e8d5b2c57a120258ed

**Comment by ajtowns** [comment]
ACK fa5eabe72117f6e3704858e8d5b2c57a120258ed

**Comment by vasild** [comment]
This silenced the compiler warnings at the time, so it was ok. I just wonder about the following reasoning, in the light of https://github.com/bitcoin/bitcoin/pull/24931:

> Remove negative lock annotations from globals ... They only make sense for mutexes that are private members.

Why is that? For example:

```cpp
Mutex g_mutex;

void f1()
{
    LOCK(g_mutex);
}

void f2()
{
    LOCK(g_mutex);
    f1();
}
```

This is undefined behavior that would be detected if `f1()` is annotated with `EXCLUSIVE_LOCKS_REQUIRED(!g_mutex)`.

To me it looks like that negative lock annotations "make sense" regardless of the scope of the mutex, no?

**Comment by MarcoFalke** [comment]
This pull was done in light of https://github.com/bitcoin/bitcoin/pull/20272#discussion_r515215137 . If https://github.com/bitcoin/bitcoin/pull/24931 fixes the mentioned issues, then that is fine.

**Comment by ajtowns** [comment]
@vasild The problem is when you call `f1` -- the negative annotation needs to keep getting propogated up the call stack:

```c++

Mutex g_mutex;
void f1() EXCLUSIVE_LOCKS_REQUIRED(!g_mutex)
{
    LOCK(g_mutex);
}

class X {
private:
    Mutex m_mut;
public:
    void f1() EXCLUSIVE_LOCKS_REQUIRED(!m_mut)
    {
        LOCK(m_mut);
    }
};
X g_x;

void caller() EXCLUSIVE_LOCKS_REQUIRED(!g_mutex)
{
    f1();
    g_x.f1();
}
```

**Comment by vasild** [comment]
Yes, it has to be propagated, but I wouldn't call it a "problem" though because it works exactly as intended. True that it creates inconvenience with code isolation (like this code should not know about that mutex), but after all a global mutex is not isolated, so IMO it is ok and even desirable to propagate its negative annotation.
